src/huggingface_hub/repocard.py

```diff
@@ -99,3 +101,117 @@ def metadata_eval_result(
         model_index, dict_factory=lambda x: {k: v for (k, v) in x if v is not None}
     )
     return {"model-index": [data]}
+
+
+def metadata_update(
+    repo_id: str,
+    metadata: Dict,
+    repo_type: str = None,
+    overwrite: bool = False,
+    token: str = None,
+) -> None:
+    """
+    Updates the metadata in the README.md of a repository on the Hugging Face Hub.
+
+    Args:
+        repo_id (`str`):
+            The name of the repository.
+        metadata (`dict`):
+            A dictionary containing the metadata to be updated.
+        repo_type (`str`, *optional*):
+            Set to `"dataset"` or `"space"` if updating to a dataset or space,
+            `None` or `"model"` if updating to a model. Default is `None`.
+        overwrite (`bool`, *optional*):
+            If set to `True` an existing field can be overwritten, otherwise
+            attempting to overwrite an existing field will cause an error.
+            Default is `False`.
+        token (`str`, *optional*):
+            The Hugging Face authentication token
+    """
+
+    filepath = hf_hub_download(
+        repo_id,
+        filename="README.md",
+        repo_type=repo_type,
+        use_auth_token=token,
+        force_download=True,
+    )
+    existing_metadata = metadata_load(filepath)
+
+    for key in metadata:
+        # update all fields except model index
+        if not key == "model-index":
+            if key in existing_metadata and not overwrite:
+                if existing_metadata[key] != metadata[key]:
+                    raise ValueError(
+                        f"""You passed a new value for the existing meta data field '{key}'. Set `overwrite=True` to overwrite existing metadata."""
+                    )
+            else:
+                existing_metadata[key] = metadata[key]
+        # update model index containing the evaluation results
+        else:
+            if "model-index" not in existing_metadata:
+                existing_metadata["model-index"] = metadata["model-index"]
+            else:
+                existing_metadata["model-index"][0][
+                    "results"
+                ] = _update_metadata_model_index(
+                    existing_metadata["model-index"][0]["results"],
+                    metadata["model-index"][0]["results"],
+                    overwrite=overwrite,
+                )
+
+    # save and push to hub
+    metadata_save(filepath, existing_metadata)
+
+    HfApi().upload_file(
+        path_or_fileobj=filepath,
+        path_in_repo="README.md",
+        repo_id=repo_id,
+        repo_type=repo_type,
+        identical_ok=True,
+        token=token,
+    )
+
+
+def _update_metadata_model_index(existing_results, new_results, overwrite=False):
```

Comment:
to better review this function, and to be able to maintain it nicely in the future, we need:
- add documentation on what the `results` look like, and how they're updated
- probably move the third and fourth `for` loops into their own function

as a reviewer, it would be nice if I could see an example covering all the control branches in this function.

Reply:
That's a good idea, I'll refactor the function. I'll also make an example for each case.